### PR TITLE
Pool balance retrieval in chainstate

### PR DESCRIPTION
**Pull request by ImplOfAnImpl**
The function `get_stake_pool_balances_at_heights` has been added to chainstate, which obtains balances of the specified pools at each height in the specified range.

### CODE REVIEW

File: chainstate/src/detail/chainstateref/mod.rs [hunk 1]
```diff
@@ -987,6 +994,107 @@ impl<'a, S: BlockchainStorageRead, V: TransactionVerificationStrategy> Chainstat
...
+            |disconnected_block_index, tx_verifier, _| -> Result<_, BlockError> {
+                let cur_height = disconnected_block_index
+                    .block_height()
+                    .prev_height()
+                    .expect("Genesis can't be disconnected");
+                assert!(cur_height >= min_height);
+
+                let pos_db = PoSAccountingDB::<_, TipStorageTag>::new(&self.db_tx);
+                let pos_delta =
+                    PoSAccountingDeltaRef::new(&pos_db, tx_verifier.accounting_delta_data());
```
Comment (azarovh) [inline chainstate/src/detail/chainstateref/mod.rs:1053]: I think you don't need this at all. `TransactionVerifier` already implements `PoSAccountingView`

Comment (ImplOfAnImpl) [reply chainstate/src/detail/chainstateref/mod.rs]: > I think you don't need this at all. TransactionVerifier already implements PoSAccountingView

Indeed, thanks. I've reverted adding that `PoSAccountingDeltaRef` stuff.

**Review by azarovh — APPROVED** [review]

File: chainstate/src/detail/chainstateref/mod.rs [hunk 2]
```diff
@@ -987,6 +992,104 @@ impl<'a, S: BlockchainStorageRead, V: TransactionVerificationStrategy> Chainstat
         Ok(block_index)
     }
 
+    pub fn get_stake_pool_balances_at_heights(
+        &self,
+        pool_ids: &[PoolId],
+        min_height: BlockHeight,
+        max_height: BlockHeight,
+    ) -> Result<BTreeMap<BlockHeight, BTreeMap<PoolId, Amount>>, BlockError> {
```
Comment (TheQuantumPhysicist) [inline chainstate/src/detail/chainstateref/mod.rs:1000]: Maybe use a range?

File: chainstate/src/detail/chainstateref/mod.rs [hunk 3]
```diff
@@ -987,6 +992,104 @@ impl<'a, S: BlockchainStorageRead, V: TransactionVerificationStrategy> Chainstat
...
+                    .block_height()
+                    .prev_height()
+                    .expect("Genesis can't be disconnected");
+                assert!(cur_height >= min_height);
+
+                let balances = Self::collect_pool_balances(pool_ids.iter(), &tx_verifier)?;
+
+                pool_ids.retain(|pool_id| {
+                    // We didn't see this pool having balance yet.
+                    pool_ids_that_had_balance.get(pool_id).is_none() ||
```
Comment (TheQuantumPhysicist) [inline chainstate/src/detail/chainstateref/mod.rs:1053]: `!pool_ids_that_had_balance.contains(pool_id)`

File: chainstate/src/detail/chainstateref/mod.rs [hunk 4]
```diff
@@ -987,6 +992,104 @@ impl<'a, S: BlockchainStorageRead, V: TransactionVerificationStrategy> Chainstat
...
+                    // We did see this pool having balance and it still does.
+                    balances.get(pool_id).is_some()
```
Comment (TheQuantumPhysicist) [inline chainstate/src/detail/chainstateref/mod.rs:1055]: balances.contains(pool_id)

**Review by TheQuantumPhysicist — APPROVED** [review]

File: chainstate/src/detail/chainstateref/mod.rs (hunk 2 above)
Comment (ImplOfAnImpl) [reply chainstate/src/detail/chainstateref/mod.rs:1000]: > Maybe use a range?

Do you insist on this?
Rust's ranges are focused around iteration but it turns out we can't iterate over a range of `BlockHeight`s (the iteration requires to implement `std::iter::Step` which is not stable yet). And outside of iteration, `Range` doesn't seem to enforce any relation between its boundaries, so it just gives some false sense of safety IMO. Also, the naming of fields/methods is confusing - both `Range` and `RangeInclusive` have `end` and you really have to look at the type to understand whether its inclusive or exclusive (I'd prefer to give specific names to inclusive range boundaries, like "min/max" or "first/last", but not "start/end").

**Review by azarovh — APPROVED** [review]

**Review by TheQuantumPhysicist — APPROVED** [review]

File: chainstate/src/detail/chainstateref/mod.rs (hunk 2 above)
Comment (ImplOfAnImpl) [reply chainstate/src/detail/chainstateref/mod.rs]: > Maybe use a range?

[discussed] Leaving this as is.

**Review by iljakuklic — APPROVED** [review]

**Review by azarovh — APPROVED** [review]

**Pull request merged**